File: problem-097/solution.py
from collections import defaultdict
from typing import DefaultDict, List
from bisect import bisect_left


class TimeMap:
    def __init__(self) -> None:
        self.keys: List[int] = []
        self.values: List[int] = []

    def set(self, key: int, value: int) -> None:
        i = bisect_left(self.keys, key)
        if i == len(self.keys):
            self.keys.append(key)
            self.values.append(value)
        elif key == self.keys[i]:
            self.values[i] = value
        else:
            self.keys.insert(i, key)
            self.values.insert(i, value)

    def get(self, key: int) -> int:
        if self.keys is None:
            return None
        i = bisect_left(self.keys, key)
        if len(self.keys) == i:
            return self.values[-1]
        elif self.keys[i] == i:
            return self.values[i]
        elif i == 0:
            return None
        else:
            return self.values[i-1]


class MultiTimeMap:
    def __init__(self) -> None:
        self.time_maps: DefaultDict[int, TimeMap] = defaultdict(TimeMap)

    def set(self, key: int, value: int, time: int) -> None:
        self.time_maps[key].set(time, value)

    def get(self, key: int, time: int) -> int:
        return self.time_maps[key].get(time)


multi_time_map = MultiTimeMap()
multi_time_map.set(1, 1, 0)
multi_time_map.set(1, 2, 2)
assert multi_time_map.get(1, 1) == 1
assert multi_time_map.get(1, 3) == 2

multi_time_map = MultiTimeMap()
multi_time_map.set(1, 1, 5)
multi_time_map.get(1, 0) is None
assert multi_time_map.get(1, 10) == 1

multi_time_map = MultiTimeMap()
multi_time_map.set(1, 1, 0)
multi_time_map.set(1, 2, 0)
assert multi_time_map.get(1, 0) == 2

File: problem-097/solution.go
package problem097

// multiTimeMap is a collection of timeMaps.
type multiTimeMap struct {
	timeMaps map[int]*timeMap
}

// newMultiTimeMap creates a new multiTimeMap.
func newMultiTimeMap() *multiTimeMap {
	m := &multiTimeMap{
		timeMaps: make(map[int]*timeMap),
	}
	return m
}

// get gets most recent value of a key relative to given time.
func (m *multiTimeMap) get(key, time int) int {
	tm := m.timeMaps[key]
	if tm == nil {
		return -1
	}
	return tm.get(time)
}

// set sets a value of a key on specific time.
func (m *multiTimeMap) set(key, val, time int) {
	if _, ok := m.timeMaps[key]; !ok {
		m.timeMaps[key] = &timeMap{}
	}
	m.timeMaps[key].set(time, val)
}

// timeMap is a map that can retrieve the value of a key at a particular time.
type timeMap struct {
	keys   []int
	values []int
}

// set sets a value given time as key
func (tm *timeMap) set(key, val int) {
	if tm.keys == nil {
		tm.keys = append(tm.keys, key)
		tm.values = append(tm.values, val)
	}

	i := getNearestKey(tm.keys, key)
	if tm.keys[i] == key {
		tm.values[i] = val
	} else {
		insert(tm.keys, i+1, key)
		insert(tm.values, i+1, val)
	}
}

// get gets a value at a particular time.
func (tm *timeMap) get(key int) int {
	i := getNearestKey(tm.keys, key)
	if i < 0 {
		return -1
	}

	return tm.values[i]
}

// getNearestKey returns the index of the nearest time key.
func getNearestKey(nums []int, key int) int {
	pos := -1
	for i, num := range nums {
		if num <= key {
			pos = i
		}
	}
	return pos
}

// insert inserts value to given int slice at given index
func insert(nums []int, index, value int) {
	if len(nums) == index {
		nums = append(nums, value)
	}
	nums = append(nums[:index+1], nums[index:]...)
	nums[index] = value
}
